Source/Dafny/Verifier/Translator.cs

```diff
@@ -11738,7 +11738,12 @@ public ForceCheckToken(IToken tok)
         Type localType = rhsTypeConstraint;  // this is a type that is appropriate for capturing the value of the RHS
         var ty = TrType(localType);
         var nm = CurrentIdGenerator.FreshId("$rhs#");
-        Bpl.Expr wh = GetWhereClause(tok, new Bpl.IdentifierExpr(tok, nm, ty), localType, etran, NOALLOC);
+        Bpl.Expr wh;
+        if (rhs is HavocRhs && localType.IsNonempty) {
```

Comment:
I argued in the issue #1619 that such temporary variables never need a `where` clause. I was thinking that such a temporary (call it `$rhs`) always gets used in the following way in the Boogie code:

```
$rhs := ...;
...
lhs := $rhs;
```

In this case, there is no reason for `$rhs` to have a `where` clause. However, I had then forgotten that for a Dafny assignment `lhs := *;` (that is, assigning an arbitrary value of the left-hand side), the generated Boogie is

```
havoc $rhs;
...
lhs := $rhs;
```

To assume the type information, one can then either follow the `havoc` with an `assume` statement or add the type information in the `where` clause of `$rhs`. But there's one more snag: if the type is possibly empty, then the `lhs := *;` assignment does not count in Dafny's definite-assignment rules. That's why it's still necessary to check `localType.IsNonempty`.

Reply:
> But there's one more snag: if the type is possibly empty, then the lhs := *; assignment does not count in Dafny's definite-assignment rules.

Can you link to the relevant code?

Reply:
You mean the code in Translator? Essentially, this means that `lhs := *;` does _not_ do a `defass$lhs := true;` assignment. In particular, `lhs := *;` translates to

```
havoc lhs;
```

and not to

```
defass$lhs := true;
havoc lhs;
```

(and similarly for multi-assignments, via a `$rhs` temporary).

Since this is (correctly) _not_ done, I have no code to point to in Translator.

There are tests for this in the new `git-issue-1619.dfy`, near the end. For example,

``` dafny
method G0() {
  var m: MaybeEmpty, other: int := *, *;
  assume G(m); // error: use before definition
```